dwh/query_lib_compatibility_tests.py

```diff
@@ -0,0 +1,244 @@
+            "num_obs",
+            "min_value",
+            "max_value",
+            "min_date",
+            "max_date",
+            "last_value",
+            "first_value",
+            "last_value_code",
+            "first_value_code",
+        }
+
+        self.assertSetEqual(set(actual_df.columns), expected_columns)
+
+    def test_obs_basic_query_include_all_other_codes(self):
+
+        end_date = "2016-01-01"
+        start_date = "2000-01-01"
+
+        # Creating a new `patient_query` to drop all previous constraints
+        # and recreate flat views.
+        patient_query = self._PatientQueryClass()
+        patient_query.include_all_other_codes(
+            min_time=start_date, max_time=end_date
+        )
+        patient_query.include_obs_in_value_and_time_range(
+            "1111", max_time="2011-01-01"
+        )
+
+        agg_df = patient_query.get_patient_obs_view()
+        self.assertTrue(
+            agg_df[agg_df["code"] == "1111"]["max_date"].max() < "2011-01-01",
+        )
+        self.assertTrue(
+            agg_df[agg_df["code"] != "1111"]["max_date"].max() < end_date
+        )
+        self.assertTrue(
+            agg_df[agg_df["code"] != "1111"]["min_date"].min() > start_date
+        )
+
+    def test_obs_query_type_codes(self):
+        _VL_CODE = "856"  # HIV VIRAL LOAD
+        _ARV_PLAN = "1255"  # ANTIRETROVIRAL PLAN
+        end_date = "2017-01-01"
+        start_date = "2000-01-01"
+
+
+        # Creating a new `patient_query` to drop all previous constraints
+        # and recreate flat views.
+        patient_query = self._PatientQueryClass()
+        patient_query.include_obs_values_in_time_range(
+            _VL_CODE, min_time=start_date, max_time=end_date
+        )
+        patient_query.include_obs_values_in_time_range(
+            _ARV_PLAN, min_time=start_date, max_time=end_date
+        )
+
+        other_end_date = "2020-01-01"
+        other_start_date = "1998-01-01"
+        patient_query.include_all_other_codes(
+            min_time=other_start_date, max_time=other_end_date
+        )
+
+        patient_query.encounter_constraints(
+            locationId=["2131aff8-2e2a-480a-b7ab-4ac53250262b"]
+        )
+
+        agg_df = patient_query.get_patient_obs_view()
+
+        self.assertTrue(
+            agg_df[agg_df["code"] == _ARV_PLAN]["max_date"].max() < end_date
+        )
+        self.assertTrue(
+            agg_df[agg_df["code"] == _ARV_PLAN]["min_date"].min() > start_date
+        )
+
+        self.assertTrue(
+            agg_df[agg_df["code"] != _ARV_PLAN]["max_date"].max()
+            < other_end_date
+        )
+        self.assertTrue(
+            agg_df[agg_df["code"] != _ARV_PLAN]["min_date"].min()
+            > other_start_date
+        )
+
+
+class PatientQueryTestBigQuery(unittest.TestCase, _PatientQueryTest):
+    """
+    Test PatientQuery API using the concreate implemenation for BigQuery
+    """
+
+    _PatientQueryClass = partial(
```

Comment:
Instead of using `partial` and implicitly assume in the parent class (`_PatientQueryTest`) that the subclass will define a `_PatientQueryClass` attribute, please use an abstract method. So in `_PatientQueryTest` there would be a base `_get_patient_query()` that returns a `PatientQuery` but its default implementation is to return `None`. Each subclass is expected to implement this base method. Also instead of calling it at the beginning of each test method, add a `setUp` method to the base class which calls this method and assign a `self._pq` which can be used in each test.

Reply:
Good idea. I did that. But, with the current implementation we cannot create this instance in `setUp` and reuse it. Note, that some of the attributes we set on `patient_query` class is stateful and therefore, we cannot always use the previous instance in tests.

Reply:
Some of the comment threads are gone because of the commit overwrite issue, so replying separately:
I believe after unifying the base class interface (e.g., resolving the `base_url` issue) this is doable. When you create an instance in [setUp](https://docs.python.org/3/library/unittest.html#unittest.TestCase.setUp) it is per test function so it is okay if it is stateful. Or did I misunderstood what you mean?

Reply:
Yes.